File: 202-happy-number/happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        bucket = set() # Keep track of the calculated sumOfSquares

        def sumOfSqr(num: int) -> int:
            digits = []
            sum = 0
            while num > 0:
                lastDigit = num % 10
                sum += lastDigit **2
                num = num // 10
            
            return sum
        
        curSumOfSq = n
        bucket.add(n)

        while curSumOfSq != 1: #Trying to reach a happy number
            curSumOfSq = sumOfSqr(curSumOfSq)
            if curSumOfSq in bucket:
                return False
            else:
                bucket.add(curSumOfSq)
        
        return True

File: 202-happy-number/happy-number.dart
class Solution {
  final isSeen = Set();

  int sumOfSquares(int n){
    int sum = 0;
    while (n>0){
      int lastDigit = n %10;
      sum += lastDigit * lastDigit;
      n = n ~/ 10;
    }

    return sum;
  }

  bool isHappy(int n) {
    isSeen.add(n);
    int curNumber = n;

    while(curNumber != 1){
      //karelerini toplave yeni curNumber
      curNumber = sumOfSquares(curNumber);

      // Dongu kontorlu yapalim
      if(isSeen.contains(curNumber)){
        return false;
      }{
        isSeen.add(curNumber);
      }

    }

    return true;
    
  }
}